From 80417356f04c5ee1cd6f636e9b043db3f2de24f2 Mon Sep 17 00:00:00 2001
From: Cheng Lian <lian@databricks.com>
Date: Tue, 3 Feb 2015 12:53:37 -0800
Subject: [PATCH] PARQUET-173: Fixes `StatisticsFilter` for `And` filter
 predicate

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/apache/incubator-parquet-mr/108)
<!-- Reviewable:end -->

Author: Cheng Lian <lian@databricks.com>

Closes #108 from liancheng/PARQUET-173 and squashes the following commits:

d188f0b [Cheng Lian] Fixes test case
be2c8a1 [Cheng Lian] Fixes `StatisticsFilter` for `And` filter predicate
---
 .../parquet/filter2/statisticslevel/StatisticsFilter.java   | 6 +++++-
 .../filter2/statisticslevel/TestStatisticsFilter.java       | 4 ++--
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/parquet-hadoop/src/main/java/parquet/filter2/statisticslevel/StatisticsFilter.java b/parquet-hadoop/src/main/java/parquet/filter2/statisticslevel/StatisticsFilter.java
index f7eea4165f..d6a484c2f8 100644
--- a/parquet-hadoop/src/main/java/parquet/filter2/statisticslevel/StatisticsFilter.java
+++ b/parquet-hadoop/src/main/java/parquet/filter2/statisticslevel/StatisticsFilter.java
@@ -242,7 +242,11 @@ public <T extends Comparable<T>> Boolean visit(GtEq<T> gtEq) {
 
   @Override
   public Boolean visit(And and) {
-    return and.getLeft().accept(this) && and.getRight().accept(this);
+    // seems unintuitive to put an || not an && here but we can
+    // drop a chunk of records if we know that either the left or
+    // the right predicate agrees that no matter what we don't
+    // need this chunk.
+    return and.getLeft().accept(this) || and.getRight().accept(this);
   }
 
   @Override
diff --git a/parquet-hadoop/src/test/java/parquet/filter2/statisticslevel/TestStatisticsFilter.java b/parquet-hadoop/src/test/java/parquet/filter2/statisticslevel/TestStatisticsFilter.java
index 03502d2bfc..d558f9b346 100644
--- a/parquet-hadoop/src/test/java/parquet/filter2/statisticslevel/TestStatisticsFilter.java
+++ b/parquet-hadoop/src/test/java/parquet/filter2/statisticslevel/TestStatisticsFilter.java
@@ -222,8 +222,8 @@ public void testAnd() {
     FilterPredicate yes = eq(intColumn, 9);
     FilterPredicate no = eq(doubleColumn, 50D);
     assertTrue(canDrop(and(yes, yes), columnMetas));
-    assertFalse(canDrop(and(yes, no), columnMetas));
-    assertFalse(canDrop(and(no, yes), columnMetas));
+    assertTrue(canDrop(and(yes, no), columnMetas));
+    assertTrue(canDrop(and(no, yes), columnMetas));
     assertFalse(canDrop(and(no, no), columnMetas));
   }
 
